In [369]:
import pandas as pd
import numpy as np
from sklearn.metrics import mean_absolute_error as mae
from sklearn.metrics import mean_squared_error as mse 
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
pd.set_option('display.max_colwidth', None)
pd.set_option('display.max_columns', 10000)
sns.set(rc={'figure.figsize':(30,12)}) # set figure size

In [206]:
df = pd.read_csv('survey_results_public.csv')

In [100]:
df

Respondent                                                MainBranch  \
0               1                            I am a developer by profession   
1               2                            I am a developer by profession   
2               3                               I code primarily as a hobby   
3               4                            I am a developer by profession   
4               5  I used to be a developer by profession, but no longer am   
...           ...                                                       ...   
64456       64858                                                       NaN   
64457       64867                                                       NaN   
64458       64898                                                       NaN   
64459       64925                                                       NaN   
64460       65112                                                       NaN   

      Hobbyist   Age Age1stCode CompFreq  CompTotal  ConvertedComp  \
0          Yes   NaN         13  Monthly        NaN            NaN   
1           No   NaN         19      NaN        NaN            NaN   
2          Yes   NaN         15      NaN        NaN            NaN   
3          Yes  25.0         18      NaN        NaN            NaN   
4          Yes  31.0         16      NaN        NaN            NaN   
...        ...   ...        ...      ...        ...            ...   
64456      Yes   NaN         16      NaN        NaN            NaN   
64457      Yes   NaN        NaN      NaN        NaN            NaN   
64458      Yes   NaN        NaN      NaN        NaN            NaN   
64459      Yes   NaN        NaN      NaN        NaN            NaN   
64460      Yes   NaN        NaN      NaN        NaN            NaN   

                  Country    CurrencyDesc CurrencySymbol  \
0                 Germany   European Euro            EUR   
1          United Kingdom  Pound sterling            GBP   
2      Russian Federation             NaN            NaN   
3                 Albania    Albanian lek            ALL   
4           United States             NaN            NaN   
...                   ...             ...            ...   
64456       United States             NaN            NaN   
64457             Morocco             NaN            NaN   
64458            Viet Nam             NaN            NaN   
64459              Poland             NaN            NaN   
64460               Spain             NaN            NaN   

                                                                                                                      DatabaseDesireNextYear  \
0                                                                                                                       Microsoft SQL Server   
1                                                                                                                                        NaN   
2                                                                                                                                        NaN   
3                                                                                                                                        NaN   
4                                                                                                                           MySQL;PostgreSQL   
...                                                                                                                                      ...   
64456                                                                                                                                    NaN   
64457  Cassandra;Couchbase;DynamoDB;Elasticsearch;Firebase;IBM DB2;MariaDB;Microsoft SQL Server;MongoDB;MySQL;Oracle;PostgreSQL;Redis;SQLite   
64458                                                                                                                                    NaN   
64459                                                   

# Understanding the types of the dataset

In [5]:
df.dtypes

Respondent              int64
MainBranch             object
Hobbyist               object
Age                   float64
Age1stCode             object
                       ...   
WebframeWorkedWith     object
WelcomeChange          object
WorkWeekHrs           float64
YearsCode              object
YearsCodePro           object
Length: 61, dtype: object

In [6]:
import matplotlib.pyplot as plt


In [9]:
# applying get_dtype_counts() function 
df.dtypes.value_counts()


object     56
float64     4
int64       1
dtype: int64

# What's the most commom undergradMajor of the participants?

In [275]:
df['UndergradMajor'].value_counts().reset_index()

index  \
0                  Computer science, computer engineering, or software engineering   
1     Another engineering discipline (such as civil, electrical, mechanical, etc.)   
2            Information systems, information technology, or system administration   
3                    A natural science (such as biology, chemistry, physics, etc.)   
4                                                        Mathematics or statistics   
5                                                    Web development or web design   
6             A business discipline (such as accounting, finance, marketing, etc.)   
7          A humanities discipline (such as literature, history, philosophy, etc.)   
8     A social science (such as anthropology, psychology, political science, etc.)   
9   Fine arts or performing arts (such as graphic design, music, studio art, etc.)   
10                                                        I never declared a major   
11                   A health science (such as nursing, pharmacy, radiology, etc.)   

    UndergradMajor  
0            31582  
1             4754  
2             4071  
3             2205  
4             1849  
5             1789  
6             1375  
7             1007  
8              933  
9              725  
10             453  
11             252

Most of the people who took the research is from Computer Science but we have a lot of Engineers in the field of programming. In the tenth we have health science, but this area I believe it is gonna have a huge increase in the next years due the application of artificial intelligence, machine learning and programming in this field.

In [274]:
data = df['UndergradMajor'].value_counts().reset_index()[1:].sort_values("UndergradMajor", ascending=True)
fig = px.bar(data, x='UndergradMajor', y='index')
fig.show()

# What it is the age that people first code?

In [34]:
df['Age1stCode'].value_counts().reset_index()

index  Age1stCode
0     15        5618
1     14        5514
2     16        5463
3     18        5382
4     12        5081
..   ...         ...
58    63           1
59    83           1
60    72           1
61    76           1
62    68           1

[63 rows x 2 columns]

In [60]:
df.query("Age1stCode == 'Younger than 5 years'")

Respondent                      MainBranch Hobbyist   Age  \
209           211  I am a developer by profession      Yes  34.0   
372           374  I am a developer by profession      Yes  40.0   
534           536  I am a developer by profession      Yes  36.0   
566           568  I am a developer by profession      Yes  36.0   
634           636  I am a developer by profession      Yes  28.0   
...           ...                             ...      ...   ...   
64383       52582                             NaN      Yes   1.0   
64401       56163                             NaN      Yes   NaN   
64412       58457                             NaN      Yes   NaN   
64421       59168                             NaN      Yes  17.0   
64438       62391                             NaN      Yes   NaN   

                 Age1stCode CompFreq  CompTotal  ConvertedComp        Country  \
209    Younger than 5 years   Yearly    80000.0        80000.0  United States   
372    Younger than 5 years   Yearly   180000.0       136136.0         Canada   
534    Younger than 5 years   Yearly   130000.0        98321.0         Canada   
566    Younger than 5 years   Yearly   150000.0       150000.0  United States   
634    Younger than 5 years   Yearly    98000.0        98000.0  United States   
...                     ...      ...        ...            ...            ...   
64383  Younger than 5 years      NaN        NaN            NaN      Indonesia   
64401  Younger than 5 years      NaN        NaN            NaN         Poland   
64412  Younger than 5 years      NaN        NaN            NaN       Djibouti   
64421  Younger than 5 years      NaN        NaN            NaN     Costa Rica   
64438  Younger than 5 years      NaN        NaN            NaN        Morocco   

               CurrencyDesc  ...                  SurveyEase  \
209    United States dollar  ...                        Easy   
372         Canadian dollar  ...                        Easy   
534         Canadian dollar  ...                        Easy   
566    United States dollar  ...                        Easy   
634    United States dollar  ...                        Easy   
...                     ...  ...                         ...   
64383                   NaN  ...                        Easy   
64401                   NaN  ...                        Easy   
64412                   NaN  ...                        Easy   
64421                   NaN  ...                        Easy   
64438                   NaN  ...  Neither easy nor difficult   

                SurveyLength Trans  \
209    Appropriate in length   NaN   
372    Appropriate in length    No   
534    Appropriate in length    No   
566    Appropriate in length    No   
634    Appropriate in length    No   
...                      ...   ...   
64383              Too short   NaN   
64401              Too short    No   
64412              Too short    No   
64421              Too short   NaN   
64438              Too short   NaN   

                                          UndergradMajor  \
209    A social science (such as anthropology, psycho...   
372    Computer science, computer engineering, or sof...   
534    Another engineering discipline (such as civil,...   
566                                                  NaN   
634    Another engineering discipline (such as civil,...   
...                                                  ...   
64383                                                NaN   
64401                                                NaN   
64412                                                NaN   
64421                                                NaN   
64438                                                NaN   

                                  WebframeDesireNextYear  \
209                                         Django;Flask   
372                                                  NaN   
534                                              ASP.NET   
566                           

 Some People answer started to code younger than 5 years, pretty impressive!

# Where those people live?

In [62]:
df['Country'].value_counts().reset_index()

index  Country
0       United States    12469
1               India     8403
2      United Kingdom     3896
3             Germany     3890
4              Canada     2191
..                ...      ...
178     Liechtenstein        1
179  Marshall Islands        1
180           Lesotho        1
181             Nauru        1
182           Grenada        1

[183 rows x 2 columns]

#  "You search for a coding solution online and the first result link is purple because you already visited it. How do you feel?"

In [101]:
df['NEWPurpleLink'].value_counts().reset_index()

index  NEWPurpleLink
0  Hello, old friend          28262
1        Indifferent          10005
2             Amused           8710
3            Annoyed           7826

# The majority of people is from US

In [349]:
data = df['Country'].value_counts().reset_index().sort_values("Country",ascending=False)
data

index  Country
0                         United States    12469
1                                 India     8403
2                        United Kingdom     3896
3                               Germany     3890
4                                Canada     2191
..                                  ...      ...
173                                Chad        1
172  Micronesia, Federated States of...        1
171                               Gabon        1
170                            Kiribati        1
182                             Grenada        1

[183 rows x 2 columns]

In [354]:
data = df['Country'].value_counts().reset_index().sort_values("Country",ascending=False)
fig = px.bar(data[0:6], x='Country', y='index')
fig.show()

# Ethnicity Brazil

In [365]:
data = df.query("Country == 'Brazil'")

In [366]:
data = data.groupby(["Ethnicity", "Country"]).size().reset_index(name="Count").sort_values('Count',ascending=False)


In [367]:
data

Ethnicity  \
46                                                                                                                                                  White or of European descent   
24                                                                                                                                                        Hispanic or Latino/a/x   
33                                                                                                                           Hispanic or Latino/a/x;White or of European descent   
45                                                                                                                                                                   Multiracial   
2                                                                                                                                                    Black or of African descent   
20                                                                                                                                                                    East Asian   
32                                                                                                                                            Hispanic or Latino/a/x;Multiracial   
39                                                                                                               Hispanic or Latino/a/x;White or of European descent;Multiracial   
5                                                                                                                             Black or of African descent;Hispanic or Latino/a/x   
0                                                                                                                                                                       Biracial   
51                                                                                                                                      White or of European descent;Multiracial   
9                                                                                                                 Black or of African descent;Hispanic or Latino/a/x;Multiracial   
25                                                                                                                                               Hispanic or Latino/a/x;Biracial   
41                                                                                              Indigenous (such as Native American, Pacific Islander, or Indigenous Australian)   
15                                                                                                                                       Black or of African descent;Multiracial   
37                                          Hispanic or Latino/a/x;White or of European descent;Indigenous (such as Native American, Pacific Islander, or Indigenous Australian)   
34                                                                                                                  Hispanic or Latino/a/x;White or of European descent;Biracial   
29                                                                                                                                         Hispanic or Latino/a/x;Middle Eastern   
43                                                                                                                                                                Middle Eastern   
18                         Black or of African descent;White or of European descent;Indigenous (such as Native American, Pacific Islander, or Indigenous Australian);Multiracial   
47                                                                                                                                         White or of European descent;Biracial   
28                                                           Hispanic or Latino/a/x;Indigenous (such as Native American, Pacific Islander, or Indigenous Australian);Multiracial   
14                         

In [368]:
fig = px.bar(data[0:5], x="Ethnicity", y="Count", title="Ethnicity Brazil")
fig.show()



# Percentual of people looking for a job

In [129]:
data =  df['JobSeek'].value_counts().reset_index()
data['percentual'] = data['JobSeek']/np.sum(data['JobSeek'])
data

index  JobSeek  \
0  I’m not actively looking, but I am open to new opportunities    29776   
1                  I am not interested in new job opportunities    13001   
2                               I am actively looking for a job     8950   

   percentual  
0    0.575637  
1    0.251339  
2    0.173024

Most of people keep the door open to opportunities

In [ ]:
# From people who work with python how many are 

# Column JobFactors

In [90]:
df['JobFactors'].value_counts().reset_index()

index  \
0                                Flex time or a flexible schedule;Languages, frameworks, and other technologies I’d be working with;Remote work options   
1    Languages, frameworks, and other technologies I’d be working with;Office environment or company culture;Opportunities for professional development   
2              Flex time or a flexible schedule;Languages, frameworks, and other technologies I’d be working with;Office environment or company culture   
3         Flex time or a flexible schedule;Languages, frameworks, and other technologies I’d be working with;Opportunities for professional development   
4                                     Flex time or a flexible schedule;Office environment or company culture;Opportunities for professional development   
..                                                                                                                                                  ...   
225                                                                                        Diversity of the company or organization;Remote work options   
226                                     Diversity of the company or organization;Financial performance or funding status of the company or organization   
227                                                                                        Diversity of the company or organization;Family friendliness   
228                                                              Diversity of the company or organization;Specific department or team I’d be working on   
229                                                       Diversity of the company or organization;How widely used or impactful my work output would be   

     JobFactors  
0          2777  
1          2570  
2          2508  
3          1666  
4          1650  
..          ...  
225           3  
226           2  
227           2  
228           1  
229           1  

[230 rows x 2 columns]

People would like to have a flex time and schedule to work, this is so evident in this survey. The organizations that don't try to have this factors will be behind in the market and also will loose their talents. Before pandemic a lot of organizations don't figure out that and when pandemic started they had a hard time to adapt.

# Onboarding and OrgSize

In [359]:
data = df.groupby(["OrgSize", "NEWOnboardGood"]).size().reset_index(name="Count").sort_values('OrgSize',ascending=True)
data

OrgSize  \
0                             1,000 to 4,999 employees   
1                             1,000 to 4,999 employees   
2                             1,000 to 4,999 employees   
3                                   10 to 19 employees   
4                                   10 to 19 employees   
5                                   10 to 19 employees   
6                             10,000 or more employees   
7                             10,000 or more employees   
8                             10,000 or more employees   
9                                 100 to 499 employees   
10                                100 to 499 employees   
11                                100 to 499 employees   
14                                    2 to 9 employees   
13                                    2 to 9 employees   
12                                    2 to 9 employees   
15                                  20 to 99 employees   
16                                  20 to 99 employees   
17                                  20 to 99 employees   
18                            5,000 to 9,999 employees   
19                            5,000 to 9,999 employees   
20                            5,000 to 9,999 employees   
21                                500 to 999 employees   
22                                500 to 999 employees   
23                                500 to 999 employees   
24  Just me - I am a freelancer, sole proprietor, etc.   
25  Just me - I am a freelancer, sole proprietor, etc.   
26  Just me - I am a freelancer, sole proprietor, etc.   

                  NEWOnboardGood  Count  
0                             No   1435  
1   Onboarding? What onboarding?    695  
2                            Yes   2520  
3                             No   1264  
4   Onboarding? What onboarding?   1052  
5                            Yes   1599  
6                             No   1806  
7   Onboarding? What onboarding?    771  
8                            Yes   3287  
9                             No   2383  
10  Onboarding? What onboarding?   1297  
11                           Yes   4282  
14                           Yes   1519  
13  Onboarding? What onboarding?   1520  
12                            No   1053  
15                            No   2870  
16  Onboarding? What onboarding?   1977  
17                           Yes   4281  
18                            No    523  
19  Onboarding? What onboarding?    270  
20                           Yes    932  
21                            No    830  
22  Onboarding? What onboarding?    410  
23                           Yes   1502  
24                            No    278  
25  Onboarding? What onboarding?   1032  
26                           Yes    425

In [361]:
fig = px.bar(data, x="OrgSize", y="Count", color="NEWOnboardGood", title="Wide-Form Input")
fig.show()

In [362]:
fig.write_html("onboard.html")

In [262]:
percentage = data.groupby(['OrgSize','NEWOnboardGood'])['Count'].sum().rename("count")


In [263]:
(percentage / percentage.groupby(level=0).sum()*100).to_frame()

count
OrgSize                                            NEWOnboardGood                         
1,000 to 4,999 employees                           No                            30.860215
                                                   Onboarding? What onboarding?  14.946237
                                                   Yes                           54.193548
10 to 19 employees                                 No                            32.286079
                                                   Onboarding? What onboarding?  26.871009
                                                   Yes                           40.842912
10,000 or more employees                           No                            30.798090
                                                   Onboarding? What onboarding?  13.148022
                                                   Yes                           56.053888
100 to 499 employees                               No                            29.929666
                                                   Onboarding? What onboarding?  16.289877
                                                   Yes                           53.780457
2 to 9 employees                                   No                            25.733138
                                                   Onboarding? What onboarding?  37.145650
                                                   Yes                           37.121212
20 to 99 employees                                 No                            31.441718
                                                   Onboarding? What onboarding?  21.658633
                                                   Yes                           46.899649
5,000 to 9,999 employees                           No                            30.318841
                                                   Onboarding? What onboarding?  15.652174
                                                   Yes                           54.028986
500 to 999 employees                               No                            30.269876
                                                   Onboarding? What onboarding?  14.952589
                                                   Yes                           54.777535
Just me - I am a freelancer, sole proprietor, etc. No                            16.023055
                                                   Onboarding? What onboarding?  59.481268
                                                   Yes                           24.495677

As you can see most of the employees have a good onboarding process as the company size increase. Most of the people are identified in companies with 20 to 99 employees which has 46% of onboarding satisfaction. In small companies, 2 to 9 employees, for example, we have a draw between "Onboarding? What Onboarding" and "Yes", maybe due some processes are being structured.